src/Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets/SocketConnection.cs

```diff
@@ -111,12 +114,37 @@ internal SocketConnection(Socket socket, SocketTransport transport)
                     }
                 }
 
+                _connectionContext.Abort(ex: null);
                 _input.Complete();
             }
             catch (Exception ex)
             {
-                _connectionContext.Abort(ex);
-                _input.Complete(ex);
+                Exception error = null;
+
+                if (ex is SocketException se)
+                {
+                    if (se.SocketErrorCode == SocketError.ConnectionReset)
+                    {
+                        // Connection reset
+                        error = new ConnectionResetException(ex.Message, ex);
+                    }
+                    else if (se.SocketErrorCode == SocketError.OperationAborted)
+                    {
+                        error = new TaskCanceledException("The request was aborted");
```

Comment:
Yes, I looked in the source code. If you call ReceiveAsync and the socket is already Disposed, it throws ObjectDisposedException.
If the ReceiveAsync is in-flight and the socket gets disposed, it throws SocketError.OperationAborted.
So it's very unlikely too occur, but not impossible.